sources/engine/Xenko.Engine/Engine/TransformComponent.cs

```diff
@@ -309,7 +309,8 @@ internal void UpdateWorldMatrixInternal(bool recursive)
                         scene.UpdateWorldMatrix();
                     }
 
-                    Matrix.Multiply(ref WorldMatrix, ref scene.WorldMatrix, out WorldMatrix);
+                    var curWorldMatrix = WorldMatrix;   // Must make a copy to use as the ref parameter, otherwise matrix will not be calculated correctly
```

Comment:
wow, just out of interest, was this a bug all along? do you remember how it affected the scene?
I had some transformation bugs and wonder whether it came from this.

Reply:
Surprisingly, this bug doesn't manifest in game.
I haven't rigorously tested this, but I tested a few matrix multiplication, and simple translations & scaling as the second matrix still multiply correctly.
Once you have rotation, it'll calculation incorrectly.
Looking at Game Studio, it seems you can only do translation offsets to scenes, so you can't encounter this bug through this code (as far as I can tell).

Reply:
We could perhaps assert/check in debug mode for cases like these, fix and compare pointers of refs and outs, if they are the same throw ?

Reply:
that's a pretty good idea!

Reply:
I am willing to make a commit for this (which can be accepted or rejected), though this is a heavily used method, so it might have an impact a user's debug build.
As far as I can tell, it needs to be written like this?
```
unsafe
{
    fixed (void* leftPtr = &left)
    fixed (void* rightPtr = &right)
    fixed (void* resultPtr = &result)
    {
        System.Diagnostics.Debug.Assert(leftPtr != resultPtr, "This method does not support 'left' parameter being the same as 'result' parameter.");
        System.Diagnostics.Debug.Assert(rightPtr != resultPtr, "This method does not support 'right' parameter being the same as 'result' parameter.");
    }
}
```

How do I expose this to propagate a user's build?
I think `#if DEBUG` doesn't work for user consumption?
Do I wrap it in another method but tag it `[Conditional("DEBUG")]`?

Reply:
The perf hit in debug mode is around 33% of the base cost of the function regardless of if we're using an exception or an assert and there's no elegant way to conditionally run logic based on the entry assembly's build type; it can't work properly with users debug build if we want optimal perfs in release so I don't think it's worthwhile to have this check in place in the end, sorry about that.

Reply:
@Eideren ok, that's a lot for a method that get's used all over the place.
@Basewq is the comment now in the summary or remarks? since it's important info, the programmer should see it in IntelliSense, i'd say.

Reply:
> it can't work properly with users debug build if we want optimal perfs

is `#ifdef DEBUG` not working here?

Reply:
Preprocessors are taken care of in the compilation process, the JIT doesn't take care of this kind of things so release dlls are already stripped out of any ``#debug``. I don't think that we can provide both debug and release versions within nugget packages and conditionally include one or the other depending on the application's build mode. 

Users would have to explicitly take the debug dlls to ever have these kinds of checks.

You can detect whether the entry assembly runs in debug or release mode but that's kind of a hack since we don't know whether the user intended the engine to run in debug mode based solely on the entry assembly's build mode or if he just has a weird project structure. 
We would also have to pay the price of having at least one branch in release mode, I don't think the issue here warrants it.
Perhaps once c# has function pointers we could take a look at this again.

Reply:
yes, that would require to ship the debug dlls... true. not important enough. but I would move the info into the summary. @Basewq

Reply:
Ok, promoted the note to summary tag.

Regarding the debug note, I've been reading up on the `Conditional` attribute, and I think this allows for the kind of branching you're looking for.
[eg. this example show](https://www.dotnetperls.com/conditional) methods are always included in every dll, but only runs when certain symbols are declared. `Debug.Write` is probably the best example.
[The IL do not get emitted](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.conditionalattribute?view=netframework-4.8) if the symbol isn't declared.
We could technically define a new symbol for users to opt in to get this extra validation, however looking at the code base, there isn't anything like this for public use already so this would be a completely new feature introduced (for advanced users...).
That said, let's leave this alone for the time being.